kms/uri/uri_test.go

```diff
@@ -0,0 +1,198 @@
+		rawuri string
+	}
+	tests := []struct {
+		name    string
+		args    args
+		want    *URI
+		wantErr bool
+	}{
+		{"ok", args{"yubikey:slot-id=9a"}, &URI{
+			URL:    &url.URL{Scheme: "yubikey", Opaque: "slot-id=9a"},
+			Values: url.Values{"slot-id": []string{"9a"}},
+		}, false},
+		{"ok file", args{"file:///tmp/ca.cert"}, &URI{
+			URL:    &url.URL{Scheme: "file", Path: "/tmp/ca.cert"},
+			Values: url.Values{},
+		}, false},
+		{"ok file simple", args{"file:/tmp/ca.cert"}, &URI{
+			URL:    &url.URL{Scheme: "file", Path: "/tmp/ca.cert"},
+			Values: url.Values{},
+		}, false},
+		{"ok file host", args{"file://tmp/ca.cert"}, &URI{
+			URL:    &url.URL{Scheme: "file", Host: "tmp", Path: "/ca.cert"},
+			Values: url.Values{},
+		}, false},
+		{"fail parse", args{"yubi%key:slot-id=9a"}, nil, true},
+		{"fail scheme", args{"yubikey"}, nil, true},
+		{"fail parse opaque", args{"yubikey:slot-id=%ZZ"}, nil, true},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := Parse(tt.args.rawuri)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("Parse() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("Parse() = %#v, want %v", got.URL, tt.want)
+			}
+		})
+	}
+}
+
+func TestParseWithScheme(t *testing.T) {
+	type args struct {
+		scheme string
+		rawuri string
+	}
+	tests := []struct {
+		name    string
+		args    args
+		want    *URI
+		wantErr bool
+	}{
+		{"ok", args{"yubikey", "yubikey:slot-id=9a"}, &URI{
+			URL:    &url.URL{Scheme: "yubikey", Opaque: "slot-id=9a"},
+			Values: url.Values{"slot-id": []string{"9a"}},
+		}, false},
+		{"ok file", args{"file", "file:///tmp/ca.cert"}, &URI{
+			URL:    &url.URL{Scheme: "file", Path: "/tmp/ca.cert"},
+			Values: url.Values{},
+		}, false},
+		{"fail parse", args{"yubikey", "yubikey"}, nil, true},
+		{"fail scheme", args{"yubikey", "awskms:slot-id=9a"}, nil, true},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := ParseWithScheme(tt.args.scheme, tt.args.rawuri)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("ParseWithScheme() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("ParseWithScheme() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestURI_Get(t *testing.T) {
+	mustParse := func(s string) *URI {
+		u, err := Parse(s)
+		if err != nil {
+			t.Fatal(err)
+		}
+		return u
+	}
+	type args struct {
+		key string
+	}
+	tests := []struct {
+		name string
+		uri  *URI
+		args args
+		want string
+	}{
+		{"ok", mustParse("yubikey:slot-id=9a"), args{"slot-id"}, "9a"},
+		{"ok first", mustParse("yubikey:slot-id=9a;slot-id=9b"), args{"slot-id"}, "9a"},
+		{"ok multiple", mustParse("yubikey:slot-id=9a;foo=bar"), args{"foo"}, "bar"},
+		{"fail missing", mustParse("yubikey:slot-id=9a"), args{"foo"}, ""},
+		{"fail in query", mustParse("yubikey:slot-id=9a?foo=bar"), args{"foo"}, ""},
```

Comment:
I think this needs to be allowed according to the abnf: https://www.rfc-editor.org/rfc/rfc7512.html#section-2.1

Reply:
This is actually testing if it can get the argument `foo` from the opaque string (`slot-id=9a`), the URI is properly parsed, but it cannot find the `foo` argument. 
To get that argument you will need to use `u.Query().Get("foo")`.